llvm/lib/ProfileData/InstrProfCorrelator.cpp

```diff
@@ -24,15 +24,38 @@
 
 using namespace llvm;
 
-/// Get the __llvm_prf_cnts section.
-Expected<object::SectionRef> getCountersSection(const object::ObjectFile &Obj) {
+namespace llvm {
+// Deprecated. Use -profile-correlate=debug-info.
+cl::opt<bool> DebugInfoCorrelate(
+    "debug-info-correlate",
+    cl::desc("Use debug info to correlate profiles (Deprecated). Use "
+             "-profile-correlate=debug-info instead."),
+    cl::init(false));
+
+cl::opt<InstrProfCorrelator::ProfCorrelatorKind> ProfileCorrelate(
+    "profile-correlate",
+    cl::desc("Use debug info or binary file to correlate profiles."),
+    cl::init(InstrProfCorrelator::NONE),
+    cl::values(clEnumValN(InstrProfCorrelator::NONE, "",
+                          "No profile correlation"),
+               clEnumValN(InstrProfCorrelator::DEBUG_INFO, "debug-info",
+                          "Use debug info to correlate")));
```

Comment:
This patch moves these options from `InstrProfiling.cpp` to `InstrProfCorrelator.cpp`, but this file doesn't actually use these flags. Is there any reason why we can't keep the flags in `InstrProfiling.cpp`?

Reply:
Do you mean Keeping DebugInfoCorrelate option can cause the linker error?

Reply:
Using extern decl of `ProfileCorrelate` in `TargetLoweringObjectFileImpl.cpp` doesn't work for binary correlation when linking `unittests/Target/PowerPC/PowerPCTests` because `InstrProfiling.o` which contains the definition is not linked for PowerPCTests.

Reply:
so  keeping the option in InstrProfCorrelator.cpp works fine without linker error? That works for me.

Reply:
What if we move it to `PGOInstrumentation.cpp`?

Reply:
> so keeping the option in InstrProfCorrelator.cpp works fine without linker error? That works for me.
Yes, PowerPCTests will linked successfully.

> What if we move it to `PGOInstrumentation.cpp`?
It still doesn't work. I guess the flags needs to be in one of those files: https://github.com/llvm/llvm-project/blob/7e34ee5291b935a1e37cd1b132552187d5c784d0/llvm/lib/ProfileData/CMakeLists.txt#L1

Reply:
I guess the root of the problem is that `Instrumentation` depends on `ProfileData`, so we can't have `InstrProfCorrelator.cpp` use a symbol defined in `Instrumentation`. That is unfortunate because `-debug-info-correlate` (`-profile-correlate=debug-info`) is a codegen flag while `InstrProfCorrelator.cpp` is primarily used by `llvm-prodata` which will not use `-debug-info-correlate`.

I really think `-debug-info-correlate` does not belong in `InstrProfCorrelator.cpp`. I'm fine with keeping `InstrProfCorrelator::ProfCorrelatorKind` in this file, but the the flags should stay in `InstrProfiling.cpp` and if needed we can pass `ProfCorrelatorKind` as a parameter.

Reply:
> I guess the root of the problem is that `Instrumentation` depends on `ProfileData`, so we can't have `InstrProfCorrelator.cpp` use a symbol defined in `Instrumentation`. 

Probably I didn't make it clear. We can keep `-debug-info-correlate` in `InstrProfiling.cpp`. 

In my [PR](https://github.com/llvm/llvm-project/pull/69493/files#diff-db16d50e8b3efb345f2adca0c4071e8bc1e06dfea64386454586a436d887ecefR484), I added a check for `-profile-correlate=binary` in `TargetLoweringObjectFileImpl.cpp`: , to determine if we should mark profile name & data sections as strippable, which requires `-profile-correlate` flag being defined in `TargetLoweringObjectFileImpl.cpp` in order to linked successfully with other targets that doesn't use `InstrProfiling.cpp`. That makes two flags separated in two files, so I moved them together to `InstrProfCorrelator.cpp`.

> That is unfortunate because `-debug-info-correlate` (`-profile-correlate=debug-info`) is a codegen flag while `InstrProfCorrelator.cpp` is primarily used by `llvm-prodata` which will not use `-debug-info-correlate`.
> I really think `-debug-info-correlate` does not belong in `InstrProfCorrelator.cpp`. I'm fine with keeping `InstrProfCorrelator::ProfCorrelatorKind` in this file, but the the flags should stay in `InstrProfiling.cpp` and if needed we can pass `ProfCorrelatorKind` as a parameter.

I didn't thought about that before. I agree it makes less sense to put codegen flags in `InstrProfCorrelator.cpp`.  I think this can be avoided by moving this flag to llvm/lib/CodeGen/CommandFlags.cpp. I'll take a look later.